lib/gooddata/lcm/actions/synchronize_user_filters.rb

```diff
@@ -133,19 +136,23 @@ def call(params)
             end
 
             filters_to_load = GoodData::UserFilterBuilder.get_filters(filters, symbolized_config)
-            puts "Synchronizing #{filters_to_load.count} filters"
+            params.gdc_logger.info("Synchronizing #{filters_to_load.count} filters")
             project.add_data_permissions(filters_to_load, run_params)
           when 'sync_multiple_projects_based_on_custom_id'
             CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               filters << row.to_hash
             end
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            filters.group_by { |u| u[multiple_projects_column] }.flat_pmap do |client_id, new_filters|
+              semaphore.synchronize { params.gdc_logger.debug("Client id #{client_id} with input data #{new_filters}") }
               fail "Client id cannot be empty" if client_id.blank?
-              project = domain.clients(client_id).project
-              fail "Client #{client_id} does not have project." unless project
-              filters_to_load = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
-              puts "Synchronizing #{filters_to_load.count} filters in project #{project.pid} of client #{client_id}"
-              project.add_data_permissions(filters_to_load, run_params)
+              client_project = domain.clients(client_id).project
+              fail "Client #{client_id} does not have project." unless client_project
+              Thread.current['filters_to_load'] = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
+              semaphore.synchronize do
+                params.gdc_logger.debug("Filters to load: #{Thread.current['filters_to_load']}")
```

Comment:
Identical code found in synchronize_user_filters.rb:152, synchronize_user_filters.rb:182 (mass = 68)